pyls/plugins/symbols.py

```diff
@@ -9,18 +9,47 @@
 @hookimpl
 def pyls_document_symbols(config, document):
     all_scopes = config.plugin_settings('jedi_symbols').get('all_scopes', True)
-    definitions = document.jedi_names(all_scopes=all_scopes)
-    return [{
-        'name': d.name,
-        'kind': _kind(d),
-        'location': {'uri': document.uri, 'range': _range(d)}
-    } for d in definitions]
+    names = document.jedi_names(all_scopes=all_scopes)
+    return list(filter(None, [serialize_name(n, document.uri) for n in names]))
 
 
-def _range(d):
+def serialize_name(name, uri):
+    result = {}
+    parent = None
+
+    try:
+        # Jedi sometimes fails here.
+        parent = name.parent()
+    except:
+        pass
+
+    # - Only include global assignments in outline.
+    # - Don't include underscore assignments in outline.
+    # - Don't include function parameters in outline.
+    if (name.type == 'statement' and
+        (name.name == '_' or parent is None or
+         (parent and parent.parent()))) or name.type == 'param':
+        return None
+    else:
+        result['name'] = name.name
+
+    result['location'] = {'uri': uri, 'range': _range(name)}
+    result['kind'] = _kind(name)
+
+    # All assignments have the filename as a there toplevel parent.
+    if parent and parent.parent():
+        result['containerName'] = parent.name
```

Comment:
Ultimately this is the crucial line of this PR right?

Reply:
This in combination with the correct end positions of the definitions is the key here.

Reply:
Ah, is that the fix to the `_range` function?

Reply:
Yes, this came mostly from the discussion at https://github.com/atom/atom-languageclient/issues/116